src/main/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStorage.java

```diff
@@ -0,0 +1,169 @@
+package org.jenkinsci.plugins.pipeline.milestone;
+
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.Extension;
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import hudson.model.Action;
+import hudson.model.Executor;
+import hudson.model.Job;
+import hudson.model.Result;
+import hudson.model.Run;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * Manages persistence of active milestones and cancellation of older builds as needed.
+ */
+public abstract class MilestoneStorage implements ExtensionPoint {
```

Comment:
Generally prefer `interface`s for new `ExtensionPoint`s. Any `static` or `final` methods here could be moved into implementation code elsewhere (`MilestoneStepExecution`).